File: apps/hrms/hrms/hr/page/employee_leave_balance/employee_leave_balance.js
frappe.pages['employee-leave-balance'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Employee Leave Balance',
		single_column: true
	});

	frappe.employee_leave_balance = new EmployeeLeaveBalance(page);
};


frappe.pages['employee-leave-balance'].on_page_show = function(wrapper) {
	$('.standard-actions.flex').addClass('hide');
	$('.standard-actions.flex').remove();
};

class EmployeeLeaveBalance {
	constructor(page) {
		this.page = page;
		this.body = $('<div></div>').appendTo(this.page.main);
		this.filters = {};
		this.data = [];
		this.current_year = new Date().getFullYear();
		this.init();
	}

	init() {
		this.setup_filters();
		this.setup_buttons();
		this.load_data();
	}

	setup_filters() {
		// Create filter area
		this.filter_area = $(`
			<div class="container my-2 p-3" style="background: white; border-radius: 8px; box-shadow: 0 1px 3px rgba(0,0,0,0.1);">
				<div class="form-group">
					<div class="row">
						<div class="col-md-3">
							<label class= "mb-0">${__('Employee')}</label>
							<div class="employee-filter"></div>
						</div>

						<div class="col-md-3">
							<label class= "mb-0">${__('Leave Type')}</label>
							<div class="leave-type-filter"></div>
						</div>

						<div class="col-md-3">
							<label class= "mb-0">${__('Year')}</label>
							<div class="year-filter"></div>
						</div>

						<div class="col-md-3 d-flex flex-column justify-content-center">
							<div class="mt-5">
								<button class="btn btn-primary btn-apply-filters">
									<i class="fa fa-search"></i> ${__('Apply')}
								</button>

								<button class="btn btn-default btn-reset-filters ml-4">
									${__('Reset')}
								</button>
							</div>
						</div>
					</div>
				</div>
			</div>
		`).appendTo(this.body);

		// Employee filter
		this.employee_filter = frappe.ui.form.make_control({
			parent: this.filter_area.find('.employee-filter'),
			df: {
				fieldtype: 'Link',
				options: 'Employee',
				placeholder: __('Select Employee')
			},
			render_input: true
		});

		// Leave Type filter
		this.leave_type_filter = frappe.ui.form.make_control({
			parent: this.filter_area.find('.leave-type-filter'),
			df: {
				fieldtype: 'Link',
				options: 'Leave Type',
				placeholder: __('Select Leave Type')
			},
			render_input: true
		});

		// Year filter
		this.year_filter = frappe.ui.form.make_control({
			parent: this.filter_area.find('.year-filter'),
			df: {
				fieldtype: 'Select',
				options: this.get_year_options(),
				default: this.current_year.toString(),
				placeholder: __('Select Year')
			},
			render_input: true
		});

		// Set default year
		this.year_filter.set_value(this.current_year.toString());

		// Add event listeners
		this.filter_area.find('.btn-apply-filters').on('click', () => this.refresh_data());
		this.filter_area.find('.btn-reset-filters').on('click', () => this.reset_filters());
	}

	get_year_options() {
		const currentYear = new Date().getFullYear();
		const years = [];
		for (let i = currentYear - 2; i <= currentYear + 1; i++) {
			years.push(i.toString());
		}
		return years.join('\n');
	}

	setup_buttons() {
		// Add refresh button
		// this.page.set_primary_action(__('Download'), () => {
		// 	this.export_data();
		// }, 'download');

		// // Add export button
		// this.page.add_menu_item(__('Export Excel'), () => {
		// 	this.export_data();
		// }, true);

		// // Add print button
		// this.page.add_menu_item(__('Print'), () => {
		// 	this.print_report();
		// }, true);
	}

	get_filters() {
		return {
			employee: this.employee_filter.get_value(),
			leave_type: this.leave_type_filter.get_value(),
			year: this.year_filter.get_value() || this.current_year.toString()
		};
	}

	reset_filters() {
		this.employee_filter.set_value('');
		this.leave_type_filter.set_value('');
		this.year_filter.set_value(this.current_year.toString());
		this.refresh_data();
	}

	load_data() {
		this.show_loading();
		const filters = this.get_filters();
		console.log('Loading data with filters:', filters);
		
		frappe.call({
			method: "hrms.hr.page.employee_leave_balance.employee_leave_balance.get_employee_leave_balance",
			args: filters,
			callback: (response) => {
				console.log('Response received:', response);
				this.hide_loading();
				if (response.message && response.message.length > 0) {
					this.data = response.message;
					this.render_data();
				} else {
					this.show_empty_state();
				}
			},
			error: (error) => {
				this.hide_loading();
				frappe.show_alert({
					message: __('Error loading data'),
					indicator: 'red'
				});
				console.error('Error:', error);
				this.show_empty_state();
			}
		});
	}

	refresh_data() {
		this.load_data();
	}

	render_data() {
		// Clear existing content
		this.body.find('.data-container').remove();
		
		if (!this.data || this.data.length === 0) {
			this.show_empty_state();
			return;
		}

		// Group data by employee if multiple employees
		const groupedData = this.group_data_by_employee();
		
		// Create data container
		this.data_container = $('<div class="data-container"></div>').appendTo(this.body);
		
		// Render each employee's data
		Object.keys(groupedData).forEach(employeeId => {
			const employeeData = groupedData[employeeId];
			const employeeInfo = employeeData[0]; // Get employee info from first record
			
			this.render_employee_section(employeeInfo, employeeData);
		});
	}

	group_data_by_employee() {
		const grouped = {};
		this.data.forEach(record => {
			if (!grouped[record.employee]) {
				grouped[record.employee] = [];
			}
			grouped[record.employee].push(record);
		});
		return grouped;
	}

	render_employee_section(employeeInfo, leaveData) {
		const filters = this.get_filters();
		const selectedYear = filters.year || this.current_year.toString();
		
		// Create employee section
		const employeeSection = $(`
			<div class="employee-section" style="margin-bottom: 30px;">
				<div class="employee-header" style="background: #f8f9fa; padding: 15px; border-radius: 8px 8px 0 0; border-left: 4px solid #17a2b8;">
					<div class="row">
						<div class="col-sm-8">
							<h4 style="margin: 0; color: #495057;">${employeeInfo.employee_name}</h4>
							<p style="margin: 5px 0 0 0; color: #6c757d;">
								<strong>ID:</strong> ${employeeInfo.employee} | 
								<strong>Department:</strong> ${employeeInfo.department || 'N/A'} | 
								<strong>Designation:</strong> ${employeeInfo.designation || 'N/A'}
							</p>
						</div>
						<div class="col-sm-4 text-right">
							<div class="year-selector" style="background: white; padding: 8px 15px; border-radius: 4px; display: inline-block;">
								<i class="fa fa-calendar" style="color: #17a2b8;"></i>
								<strong style="margin-left: 8px; color: #495057;">${selectedYear}</strong>
							</div>
						</div>
					</div>
				</div>
				<div class="leave-table-container" style="background: white; padding: 20px; border-radius: 0 0 8px 8px; box-shadow: 0 2px 4px rgba(0,0,0,0.1);">
					<div class="table-responsive">
						<table class="table table-bordered allocated-leaves-table">
							<thead>
								<tr>
									<th>Leave Type</th>
									<th>Total Allocated Leaves</th>
									<th>Expired Leaves</th>
									<th>Used Leaves</th>
									<th>Leaves Pending Approval</th>
									<th>Available Leaves</th>
									<th>Actions</th>
								</tr>
							</thead>
							<tbody>
		`);

		leaveData.forEach(leave => {
			employeeSection.find('tbody').append(`
				<tr data-employee="${leave.employee}" data-leave-type="${leave.leave_type}">
					<td>${leave.leave_type_name}</td>
					<td>${leave.total_leaves}</td>
					<td>${leave.expired_leaves}</td>
					<td>${leave.leaves_taken}</td>
					<td>${leave.leaves_pending_approval}</td>
					<td><strong>${leave.remaining_leaves}</strong></td>
					<td>
						<button class="btn btn-sm btn-info view-details-btn" style="font-size: 12px;">
							View Details
						</button>
					</td>
				</tr>
			`);
		});

		employeeSection.find('tbody').append('</tbody></table></div></div></div>');

		// Add click handler for view details
		employeeSection.find('.view-details-btn').on('click', (e) => {
			e.stopPropagation();
			const row = $(e.target).closest('tr');
			const leave_type = row.data('leave-type');
			const leaveDataItem = leaveData.find(l => l.leave_type === leave_type);
			this.show_leave_details(leaveDataItem);
		});

		this.data_container.append(employeeSection);
	}

	show_leave_details(leaveData) {
		// Create a dialog to show detailed leave applications
		const dialog = new frappe.ui.Dialog({
			title: `${leaveData.leave_type_name} - ${leaveData.employee_name}`,
			size: 'large',
			fields: [
				{
					fieldtype: 'HTML',
					fieldname: 'leave_details_html'
				}
			]
		});

		// Prepare HTML content for leave applications - matching Allocated Leaves
		let detailsHtml = `
			<div style="margin-bottom: 20px;">
				<h5>Leave Summary</h5>
				<div class="row">
					<div class="col-sm-4">
						<strong>Total Allocated Leaves:</strong><br>
						<span style="font-size: 18px; color: #28a745;">${leaveData.total_leaves || 0}</span>
					</div>
					<div class="col-sm-2">
						<strong>Expired Leaves:</strong><br>
						<span style="font-size: 18px; color: #dc3545;">${leaveData.expired_leaves || 0}</span>
					</div>
					<div class="col-sm-2">
						<strong>Used Leaves:</strong><br>
						<span style="font-size: 18px; color: #ffc107;">${leaveData.leaves_taken || 0}</span>
					</div>
					<div class="col-sm-2">
						<strong>Pending Approval:</strong><br>
						<span style="font-size: 18px; color: #6c757d;">${leaveData.leaves_pending_approval || 0}</span>
					</div>
					<div class="col-sm-2">
						<strong>Available Leaves:</strong><br>
						<span style="font-size: 18px; color: #17a2b8;">${leaveData.remaining_leaves || 0}</span>
					</div>
				</div>
			</div>
		`;

		if (leaveData.leave_applications && leaveData.leave_applications.length > 0) {
			detailsHtml += `
				<h5>Recent Leave Applications</h5>
				<div class="table-responsive">
					<table class="table table-striped">
						<thead>
							<tr>
								<th>Application</th>
								<th>From Date</th>
								<th>To Date</th>
								<th>Days</th>
								<th>Status</th>
							</tr>
						</thead>
						<tbody>
			`;

			leaveData.leave_applications.forEach(app => {
				detailsHtml += `
					<tr>
						<td><a href="/app/leave-application/${app.name}" target="_blank">${app.name}</a></td>
						<td>${frappe.datetime.str_to_user(app.from_date)}</td>
						<td>${frappe.datetime.str_to_user(app.to_date)}</td>
						<td>${app.total_leave_days}</td>
						<td><span class="badge badge-success">${app.status}</span></td>
					</tr>
				`;
			});

			detailsHtml += `
						</tbody>
					</table>
				</div>
			`;
		} else {
			detailsHtml += `<p class="text-muted">No leave applications found for this leave type.</p>`;
		}

		dialog.fields_dict.leave_details_html.$wrapper.html(detailsHtml);
		dialog.show();
	}

	show_employee_details(employee) {
		frappe.set_route('Form', 'Employee', employee);
	}

	show_loading() {
		this.body.find('.data-container').remove();
		this.loading_container = $(`
			<div class="data-container">
				<div class="text-center" style="padding: 100px;">
					<i class="fa fa-spinner fa-spin fa-3x text-muted"></i>
					<p class="text-muted" style="margin-top: 20px; font-size: 16px;">${__('Loading leave balance data...')}</p>
				</div>
			</div>
		`).appendTo(this.body);
	}

	hide_loading() {
		if (this.loading_container) {
			this.loading_container.remove();
		}
	}

	show_empty_state() {
		this.body.find('.data-container').remove();
		$(`
			<div class="data-container">
				<div class="text-center" style="padding: 100px;">
					<i class="fa fa-calendar-times-o fa-4x text-muted"></i>
					<h4 class="text-muted" style="margin-top: 20px;">${__('No Leave Balance Data Found')}</h4>
					<p class="text-muted">${__('Try adjusting your filters or check if leave allocations exist for the selected criteria')}</p>
					<button class="btn btn-primary btn-sm" onclick="frappe.employee_leave_balance.reset_filters()">
						${__('Reset Filters')}
					</button>
				</div>
			</div>
		`).appendTo(this.body);
	}

	export_data() {
		if (!this.data || this.data.length === 0) {
			frappe.show_alert({
				message: __('No data to export'),
				indicator: 'orange'
			});
			return;
		}

		frappe.show_alert({
			message: __('Export functionality will be implemented'),
			indicator: 'blue'
		});
	}

	print_report() {
		if (!this.data || this.data.length === 0) {
			frappe.show_alert({
				message: __('No data to print'),
				indicator: 'orange'
			});
			return;
		}

		window.print();
	}
}

File: apps/hrms/hrms/hr/page/employee_leave_balance/employee_leave_balance.py
import frappe
from frappe import _
from frappe.utils import flt, cint, getdate, formatdate, nowdate, add_years
from frappe.desk.query_report import get_columns_dict
import json
from datetime import datetime
from hrms.hr.doctype.leave_application.leave_application import get_leave_details

from frappe.utils import getdate, add_days, flt

@frappe.whitelist()
def get_employee_leave_balance(employee=None, leave_type=None, year=None):
    
    if not employee:
        emp = get_employee_details(frappe.session.user)
        if emp:
            employee = emp.get("name")
            employee_name = emp.get("employee_name")
            company_email = emp.get("company_email")
        else:
            employee = None
        
        
    try:
        # Validate permissions first
        validate_permissions()
        
        filters = build_filters(employee, leave_type, year)
        data = get_leave_balance_data(filters)
        return data
        
    except frappe.ValidationError:
       
        raise
    except Exception as e:
        frappe.log_error(f"Error in get_employee_leave_balance: {str(e)}")
        frappe.logger().error(f"Detailed error: {str(e)}")
        frappe.throw(_("Error fetching leave balance data: {0}").format(str(e)))


def get_employee_details(user):
    employee = frappe.get_all(
        "Employee",
        filters=[
            ["status", "=", "Active"],
            ["company_email", "=", user]
        ],
        fields=["name", "employee_name", "company_email"],
        limit_page_length=1
    )
    return employee[0] if employee else None


def build_filters(employee, leave_type, year):
    filters = {}
    
    if employee:
        filters['employee'] = employee
    
    if leave_type:
        filters['leave_type'] = leave_type
    
   
    current_year = datetime.now().year
    filters['year'] = int(year) if year else current_year
    
    filters['from_date'] = f"{filters['year']}-01-01"
    filters['to_date'] = f"{filters['year']}-12-31"
    
    return filters

def format_leave_value(value):
    """Format leave value to show as integer or 1 decimal place."""
    value = flt(value)
    # If it's effectively an integer (like 4.0), show as 4
    if value.is_integer():
        return int(value)
    # Else show one decimal (like 4.5)
    return round(value, 1)

def get_leave_balance_data(filters):
    try:
        from_date = getdate(filters["from_date"])
        to_date = getdate(filters["to_date"])

        # Calculate the reference date for leave details (as of end of year if past, else today)
        current_year = datetime.now().year
        if filters['year'] < current_year:
            calc_date = getdate(f"{filters['year']}-12-31")
        else:
            calc_date = getdate(nowdate())

        # Get unique employees with allocations in the period
        if filters.get("employee"):
            employees = [filters["employee"]]
        else:
            employees_query = """
                SELECT DISTINCT employee 
                FROM `tabLeave Allocation` la
                WHERE la.docstatus = 1 
                AND la.from_date <= %(to_date)s 
                AND la.to_date >= %(from_date)s
            """
            emp_list = frappe.db.sql(employees_query, filters, as_list=True)
            employees = [e[0] for e in emp_list]

        final_data = []

        for emp in employees:
            try:
                # Use the same method as Leave Application dashboard, with calculated date
                leave_details_resp = get_leave_details(emp, calc_date)
                allocation = leave_details_resp.get("leave_allocation", {})

                emp_info = frappe.db.get_value(
                    "Employee", 
                    emp, 
                    ["employee_name", "designation", "department", "company"], 
                    as_dict=1
                )

                for lt, details in allocation.items():
                    if filters.get("leave_type") and lt != filters["leave_type"]:
                        continue

                    leave_type_name = frappe.db.get_value("Leave Type", lt, "leave_type_name") or lt

                    # Get leave applications for details
                    leave_applications = get_leave_application_details(emp, lt, filters)

                    row = {
                        "employee": emp,
                        "employee_name": emp_info.get("employee_name"),
                        "designation": emp_info.get("designation"),
                        "department": emp_info.get("department"),
                        "company": emp_info.get("company"),
                        "leave_type": lt,
                        "leave_type_name": leave_type_name,
                        "total_leaves": format_leave_value(details["total_leaves"]),
                        "expired_leaves": format_leave_value(details["expired_leaves"]),
                        "leaves_taken": format_leave_value(details["leaves_taken"]),
                        "leaves_pending_approval": format_leave_value(details["leaves_pending_approval"]),
                        "remaining_leaves": format_leave_value(details["remaining_leaves"]),
                        "total_applications": len(leave_applications),
                        "leave_applications": leave_applications,
                    }

                    final_data.append(row)

            except Exception as e:
                frappe.logger().error(f"[get_leave_balance_data] Error processing {emp}: {str(e)}")

        return final_data

    except Exception as e:
        frappe.logger().error(f"Error in get_leave_balance_data(): {str(e)}")
        raise

def get_leave_application_details(employee, leave_type, filters):
    """
    Get detailed leave application data for specific employee and leave type
    
    Args:
        employee (str): Employee ID
        leave_type (str): Leave Type
        filters (dict): Additional filters
    
    Returns:
        list: Leave application details
    """
    try:
        conditions = [
            "la.docstatus = 1", 
            "la.employee = %(employee)s", 
            "la.leave_type = %(leave_type)s",
            "la.status = 'Approved'"
        ]
        values = {'employee': employee, 'leave_type': leave_type}
        
        # Add date range filter
        if filters.get('from_date') and filters.get('to_date'):
            conditions.append("la.from_date >= %(from_date)s")
            conditions.append("la.to_date <= %(to_date)s")
            values['from_date'] = filters['from_date']
            values['to_date'] = filters['to_date']
        
        where_clause = " AND ".join(conditions)
        
        query = f"""
            SELECT 
                la.name,
                la.posting_date,
                la.from_date,
                la.to_date,
                la.total_leave_days,
                la.status,
                la.leave_type,
                la.description,
                la.leave_approver,
                la.follow_via_email
            FROM 
                `tabLeave Application` la
            WHERE 
                {where_clause}
            ORDER BY 
                la.from_date DESC
            
        """
        
        return frappe.db.sql(query, values, as_dict=True)
        
    except Exception as e:
        frappe.logger().error(f"Error in get_leave_application_details: {str(e)}")
        return []

def validate_permissions():
    if not frappe.has_permission('Employee', 'read'):
        frappe.throw(_("You don't have permission to access employee data"))
    
    if not frappe.has_permission('Leave Application', 'read'):
        frappe.throw(_("You don't have permission to access leave application data"))

@frappe.whitelist()
def export_leave_balance(employee=None, leave_type=None, year=None):
   
    try:
        validate_permissions()
        
        from frappe.utils.xlsxutils import make_xlsx
        
        filters = build_filters(employee, leave_type, year)
        data = get_leave_balance_data(filters)
        
        if not data:
            frappe.throw(_("No data found to export"))
        
        # Prepare data for Excel - matching Allocated Leaves
        columns = [
            _("Employee ID"),
            _("Employee Name"),
            _("Department"),
            _("Designation"),
            _("Leave Type"),
            _("Total Allocated Leaves"),
            _("Expired Leaves"),
            _("Used Leaves"),
            _("Leaves Pending Approval"),
            _("Available Leaves"),
            _("Total Applications")
        ]
        
        rows = []
        for row in data:
            rows.append([
                row.get('employee'),
                row.get('employee_name'),
                row.get('department'),
                row.get('designation'),
                row.get('leave_type_name'),
                row.get('total_leaves', 0),
                row.get('expired_leaves', 0),
                row.get('leaves_taken', 0),
                row.get('leaves_pending_approval', 0),
                row.get('remaining_leaves', 0),
                row.get('total_applications', 0)
            ])
        
        # Create Excel file
        xlsx_data = make_xlsx([columns] + rows, "Employee Leave Balance")
        
        # Save file
        filename = f"employee_leave_balance_{filters['year']}_{frappe.utils.nowdate()}.xlsx"
        
        # Create file record
        file_doc = frappe.get_doc({
            "doctype": "File",
            "file_name": filename,
            "content": xlsx_data.getvalue(),
            "is_private": 1
        })
        file_doc.insert()
        
        return {
            'file_url': file_doc.file_url,
            'filename': filename
        }
        
    except Exception as e:
        frappe.log_error(f"Error in export_leave_balance: {str(e)}")
        frappe.throw(_("Error exporting data: {0}").format(str(e)))

@frappe.whitelist()
def get_dashboard_data():
   
    try:
        validate_permissions()
        
        current_date = nowdate()
        current_year = datetime.now().year
        
        # Get total active employees
        total_employees = frappe.db.count('Employee', {'status': 'Active'})
        
        # Get employees on leave today
        employees_on_leave = frappe.db.sql("""
            SELECT COUNT(DISTINCT employee) as count
            FROM `tabLeave Application`
            WHERE status = 'Approved'
            AND docstatus = 1
            AND %s BETWEEN from_date AND to_date
        """, (current_date,), as_dict=True)[0].get('count', 0)
        
        # Get pending leave applications
        pending_applications = frappe.db.count('Leave Application', {
            'status': 'Open',
            'docstatus': 0
        })
        
        # Get employees with low leave balance (less than 5 days)
        low_balance_query = """
            SELECT COUNT(*) as count
            FROM (
                SELECT 
                    emp.name as employee,
                    lt.name as leave_type,
                    COALESCE(
                        (SELECT SUM(total_leaves_allocated) 
                         FROM `tabLeave Allocation` 
                         WHERE employee = emp.name 
                         AND leave_type = lt.name 
                         AND docstatus = 1
                         AND YEAR(from_date) = %s), 0
                    ) - COALESCE(
                        (SELECT SUM(total_leave_days)
                         FROM `tabLeave Application` 
                         WHERE employee = emp.name 
                         AND leave_type = lt.name 
                         AND docstatus = 1 
                         AND status = 'Approved'
                         AND YEAR(from_date) = %s), 0
                    ) as balance
                FROM `tabEmployee` emp
                CROSS JOIN `tabLeave Type` lt
                WHERE emp.status = 'Active'
                HAVING balance > 0 AND balance < 5
            ) as low_balance_data
        """
        
        low_balance_result = frappe.db.sql(low_balance_query, (current_year, current_year), as_dict=True)
        low_balance_employees = low_balance_result[0].get('count', 0) if low_balance_result else 0
        
        # Get leave applications this month
        monthly_applications = frappe.db.sql("""
            SELECT COUNT(*) as count
            FROM `tabLeave Application`
            WHERE MONTH(creation) = MONTH(%s)
            AND YEAR(creation) = YEAR(%s)
            AND docstatus = 1
        """, (current_date, current_date), as_dict=True)[0].get('count', 0)
        
        return {
            'total_employees': total_employees,
            'employees_on_leave': employees_on_leave,
            'pending_applications': pending_applications,
            'low_balance_employees': low_balance_employees,
            'monthly_applications': monthly_applications,
            'current_date': current_date,
            'current_year': current_year
        }
        
    except Exception as e:
        frappe.log_error(f"Error in get_dashboard_data: {str(e)}")
        return {}

@frappe.whitelist()
def get_employee_suggestions(txt):
    """
    Get employee suggestions for autocomplete
    
    Args:
        txt (str): Search text
    
    Returns:
        list: Employee suggestions
    """
    try:
        return frappe.db.sql("""
            SELECT name, employee_name
            FROM `tabEmployee`
            WHERE status = 'Active'
            AND (name LIKE %(txt)s OR employee_name LIKE %(txt)s)
            ORDER BY employee_name
            LIMIT 20
        """, {'txt': f'%{txt}%'}, as_dict=True)
        
    except Exception as e:
        frappe.logger().error(f"Error in get_employee_suggestions: {str(e)}")
        return []

@frappe.whitelist()
def get_leave_type_suggestions(txt):
    """
    Get leave type suggestions for autocomplete
    
    Args:
        txt (str): Search text
    
    Returns:
        list: Leave type suggestions
    """
    try:
        return frappe.db.sql("""
            SELECT name, leave_type_name
            FROM `tabLeave Type`
            WHERE (name LIKE %(txt)s OR leave_type_name LIKE %(txt)s)
            ORDER BY leave_type_name
            LIMIT 20
        """, {'txt': f'%{txt}%'}, as_dict=True)
        
    except Exception as e:
        frappe.logger().error(f"Error in get_leave_type_suggestions: {str(e)}")
        return []